src/lib.rs

```diff
@@ -375,46 +391,123 @@ const fn byte2hex(byte: u8, table: &[u8; 16]) -> (u8, u8) {
 /// assert_eq!(hex::encode_to_slice(b"kiwi", &mut bytes), Err(FromHexError::InvalidStringLength));
 ///
 /// // you can do this instead:
-/// hex::encode_to_slice(b"kiwi", &mut bytes[..4 * 2])?;
+/// let string = hex::encode_to_slice(b"kiwi", &mut bytes[..4 * 2])?;
+/// assert_eq!(string, "6b697769");
 /// assert_eq!(&bytes, b"6b697769\0\0");
 /// # Ok(())
 /// # }
 /// ```
-pub fn encode_to_slice<T: AsRef<[u8]>>(input: T, output: &mut [u8]) -> Result<(), FromHexError> {
-    if input.as_ref().len() * 2 != output.len() {
+#[inline]
+pub fn encode_to_slice<T: AsRef<[u8]>>(
+    input: T,
+    output: &mut [u8],
+) -> Result<&mut str, FromHexError> {
+    encode_to_slice_inner(input.as_ref(), output, HEX_CHARS_LOWER)
+}
+
+/// Encodes some bytes into the provided byte buffer, and then returns the buffer as a string.
+///
+/// Apart from the characters' casing, this works exactly like [`encode_to_slice()`].
+///
+/// The output buffer has to be able to hold exactly `input.len() * 2` bytes,
+/// or this function will return an error.
+///
+/// # Example
+///
+/// ```
+/// # use hex::FromHexError;
+/// # fn main() -> Result<(), FromHexError> {
+/// let mut bytes = [0u8; 4 * 2];
+///
+/// let string = hex::encode_to_slice_upper(b"kiwi", &mut bytes)?;
+/// assert_eq!(string, "6B697769");
+/// # Ok(())
+/// # }
+/// ```
+///
+/// If the buffer is too large, an error is returned:
+///
+/// ```
+/// use hex::FromHexError;
+/// # fn main() -> Result<(), FromHexError> {
+/// let mut bytes = [0_u8; 5 * 2];
+///
+/// assert_eq!(hex::encode_to_slice_upper(b"kiwi", &mut bytes), Err(FromHexError::InvalidStringLength));
+///
+/// // you can do this instead:
+/// let string = hex::encode_to_slice_upper(b"kiwi", &mut bytes[..4 * 2])?;
+/// assert_eq!(string, "6B697769");
+/// assert_eq!(&bytes, b"6B697769\0\0");
+/// # Ok(())
+/// # }
+/// ```
+#[inline]
+pub fn encode_to_slice_upper<T: AsRef<[u8]>>(
+    input: T,
+    output: &mut [u8],
+) -> Result<&mut str, FromHexError> {
+    encode_to_slice_inner(input.as_ref(), output, HEX_CHARS_UPPER)
+}
+
+#[inline]
+fn encode_to_slice_inner<'a>(
+    input: &[u8],
+    output: &'a mut [u8],
+    table: &[u8; 16],
+) -> Result<&'a mut str, FromHexError> {
+    if input.len() * 2 != output.len() {
         return Err(FromHexError::InvalidStringLength);
     }
 
-    for (byte, (i, j)) in input
-        .as_ref()
-        .iter()
-        .zip(generate_iter(input.as_ref().len() * 2))
-    {
-        let (high, low) = byte2hex(*byte, HEX_CHARS_LOWER);
-        output[i] = high;
-        output[j] = low;
+    for (out, byte) in output.chunks_exact_mut(2).zip(input) {
+        let (high, low) = byte2hex(*byte, table);
+        out[0] = high;
+        out[1] = low;
     }
 
-    Ok(())
+    // SAFETY: output was just fully filled with only ascii characters
+    let output = unsafe { core::str::from_utf8_unchecked_mut(output) };
```

Comment:
If input is an empty slice, then the zip is an empty iterator, and no data is written to output. Thus there is no reason to assume output contains valid utf8, though the subslice that was actually written does.